File: src/JS/copying_from_JIRA.js
function fetchTickets() {
  var jiraDomain = '';
  var username = '';
  var apiToken = '';
  var generalFilterId = '';  // Replace with your general filter ID
  var techDebtFilterId = '';  // Replace with your tech debt filter ID
  var labels = [''];  // Replace with the labels you are looking for

 // Open the specific spreadsheet using its URL
  var spreadsheetUrl = 'Google sheet URL';
  var spreadsheet = SpreadsheetApp.openByUrl(spreadsheetUrl);
  var originalSheet = spreadsheet.getSheetByName('SheetName');  // Replace with the actual name of the sheet to be copied

   if (!originalSheet) {
    throw new Error('Original sheet not found: SheetName');
  }
  
  // Copy the sheet
  var newSheet = originalSheet.copyTo(spreadsheet);
  var newSheetName = getFormattedDate(1);  // Get tomorrow's date as the sheet name
  newSheet.setName(newSheetName);

  // Move the new sheet before the original sheet
  var originalSheetIndex = spreadsheet.getSheets().indexOf(originalSheet);
  spreadsheet.setActiveSheet(newSheet);
  spreadsheet.moveActiveSheet(originalSheetIndex);

  // Clear the contents of the new sheet but preserve the header and formulas in column H
  var range = newSheet.getDataRange();
  var formulas = range.getFormulas();
  var notesFormulas = formulas.map(row => [row[7]]); // Copy formulas from column H
  
  // Clear all rows below the header row
  newSheet.getRange(2, 1, newSheet.getMaxRows()-1, newSheet.getMaxColumns()).clearContent();
  
  // Preserve formulas in column H
  newSheet.getRange(2, 8, notesFormulas.length, 1).setFormulas(notesFormulas); 
  
  // Remove strikethrough formatting from the entire sheet
  removeStrikethrough(newSheet);

  // Fetch tickets based on the general filter and labels
  fetchAndWriteTickets(newSheet, jiraDomain, username, apiToken, generalFilterId, labels, 150);

  // Fetch tech debt tickets based on the tech debt filter
  fetchAndWriteTechDebtTickets(newSheet, jiraDomain, username, apiToken, techDebtFilterId);
}

function fetchAndWriteTickets(sheet, jiraDomain, username, apiToken, filterId, labels, maxResults) {
  var labelCondition = labels.map(function(label) {
    return `labels = "${label}"`;
  }).join(' OR ');

  var jql = `
    filter = ${filterId} AND statusCategory != Done AND (${labelCondition}) AND
    (
      (sprint is EMPTY AND "Story Points[Number]" is EMPTY) OR
      (sprint is not EMPTY AND "Story Points[Number]" is not EMPTY AND created <= -60d)
    )
  `;
  var response = callJiraApi(jiraDomain, username, apiToken, jql, maxResults);

  if (response && response.issues) {
    var issues = response.issues;
    Logger.log(issues); // Log the response for debugging
    
    // Group issues by project key
    var issuesByProject = issues.reduce(function(result, issue) {
      var projectKey = issue.fields.project.key;
      if (!result[projectKey]) {
        result[projectKey] = [];
      }
      result[projectKey].push(issue);
      return result;
    }, {});

    var rows = [];
    var projectKeys = Object.keys(issuesByProject);
    for (var i = 0; i < projectKeys.length; i++) {
      var projectIssues = issuesByProject[projectKeys[i]];
      var projectRows = projectIssues.slice(0, 6).map(function(issue) {
        var createdDate = new Date(issue.fields.created);
        var formattedDate = Utilities.formatDate(createdDate, Session.getScriptTimeZone(), 'M/dd/yyyy HH:mm:ss');
        return [
          issue.key, 
          issue.fields.summary, 
          issue.fields.assignee ? issue.fields.assignee.displayName : 'Unassigned',
          issue.fields.reporter ? issue.fields.reporter.displayName : 'Unknown',
          issue.fields.priority ? issue.fields.priority.name : 'N/A',
          issue.fields.customfield_10000 ? issue.fields.customfield_10000.value : 'N/A',  // Replace with your Product field ID if different
          formattedDate,
          ''  // Placeholder for Notes column with preserved formulas
        ];
      });
      rows = rows.concat(projectRows);
    }

    if (rows.length > 0) {
      sheet.getRange(2, 1, rows.length, rows[0].length).setValues(rows);  // Set values including the Notes column
    }
  }
}

function fetchAndWriteTechDebtTickets(sheet, jiraDomain, username, apiToken, filterId) {
  var jql = `
    filter = ${filterId} AND statusCategory != Done AND
    (
      (sprint is EMPTY AND "Story Points[Number]" is EMPTY) OR
      (sprint is not EMPTY AND "Story Points[Number]" is not EMPTY AND created <= -60d)
    )
    ORDER BY created DESC
  `;
  var response = callJiraApi(jiraDomain, username, apiToken, jql, 1);

  if (response && response.issues && response.issues.length > 0) {
    var issue = response.issues[0];
    var createdDate = new Date(issue.fields.created);
    var formattedDate = Utilities.formatDate(createdDate, Session.getScriptTimeZone(), 'M/dd/yyyy HH:mm:ss');
    sheet.appendRow([
      issue.key, 
      issue.fields.summary, 
      issue.fields.assignee ? issue.fields.assignee.displayName : 'Unassigned',
      issue.fields.reporter ? issue.fields.reporter.displayName : 'Unknown',
      issue.fields.priority ? issue.fields.priority.name : 'N/A',
      issue.fields.customfield_10000 ? issue.fields.customfield_10000.value : 'N/A',  // Replace with your Product field ID if different
      formattedDate,
      ''  // Placeholder for Notes column with preserved formulas
    ]);
  }
}

function callJiraApi(jiraDomain, username, apiToken, jql, maxResults) {
  var url = `${jiraDomain}/rest/api/2/search?jql=${encodeURIComponent(jql)}&maxResults=${maxResults}`;
  var headers = {
    "Authorization": "Basic " + Utilities.base64Encode(username + ':' + apiToken),
    "Accept": "application/json"
  };

  var response = UrlFetchApp.fetch(url, {
    "method": "get",
    "headers": headers
  });

  return JSON.parse(response.getContentText());
}

function shuffleArray(array) {
  for (var i = array.length - 1; i > 0; i--) {
    var j = Math.floor(Math.random() * (i + 1));
    var temp = array[i];
    array[i] = array[j];
    array[j] = temp;
  }
  return array;
}

function getFormattedDate(daysToAdd) {
  var date = new Date();
  date.setDate(date.getDate() + daysToAdd);
  var monthNames = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];
  var day = date.getDate();
  var month = monthNames[date.getMonth()];
  var year = date.getFullYear();
  return month + '-' + day + '-' + year;
}

function applyConditionalFormatting(sourceSheet, targetSheet) {
  var rules = sourceSheet.getConditionalFormatRules();
  targetSheet.setConditionalFormatRules(rules);
}

function removeStrikethrough(sheet) {
  var range = sheet.getDataRange();
  var richTextValues = range.getRichTextValues();
  var newRichTextValues = [];

  for (var i = 0; i < richTextValues.length; i++) {
    var row = [];
    for (var j = 0; j < richTextValues[i].length; j++) {
      var cell = richTextValues[i][j];
      if (cell) {
        var newTextStyle = cell.getTextStyle().copy().setStrikethrough(false).build();
        var newCell = SpreadsheetApp.newRichTextValue()
          .setText(cell.getText())
          .setTextStyle(newTextStyle)
          .build();
        row.push(newCell);
      } else {
        row.push(cell);
      }
    }
    newRichTextValues.push(row);
  }
  range.setRichTextValues(newRichTextValues);
}

File: src/Python/copying_from_JIRA.py
import requests
from jira import JIRA
import gspread
from oauth2client.service_account import ServiceAccountCredentials
import pandas as pd
from datetime import datetime, timedelta

# 1. Configuration (Store in a separate `config.py` file and import)
jira_domain = ''
username = ''
api_token = ''
general_filter_id = '' 
tech_debt_filter_id = ''
labels = [''] 
spreadsheet_url = 'Google sheet URL'
sheet_name = 'SheetName'

# 2. Google Sheets Authentication
scope = ['https://spreadsheets.google.com/feeds', 'https://www.googleapis.com/auth/drive']
creds = ServiceAccountCredentials.from_json_keyfile_name('your_credentials.json', scope) # Replace with your credentials file
client = gspread.authorize(creds)

# 3. Open Spreadsheet and Create New Sheet
spreadsheet = client.open_by_url(spreadsheet_url)
original_sheet = spreadsheet.worksheet(sheet_name)

tomorrow = datetime.now() + timedelta(days=1)
new_sheet_name = tomorrow.strftime('%b-%d-%Y')
new_sheet = original_sheet.duplicate(new_sheet_name=new_sheet_name)

# 4. Clear Data and Preserve Formulas
all_data = new_sheet.get_all_values()
formulas = [row[7] for row in all_data]  # Assuming formulas are in column H (index 7)

# Clear data but keep header row
new_sheet.clear()
new_sheet.append_row(all_data[0])  # Append the header row back

# Re-apply formulas to column H
for i, formula in enumerate(formulas[1:]):  # Skip header row
    new_sheet.update_cell(i + 2, 8, formula)  # Start from row 2

# 5. Fetch and Write Jira Tickets
def fetch_and_write_tickets(sheet, jira_domain, username, api_token, filter_id, labels, max_results):
    label_condition = ' OR '.join([f'labels = "{label}"' for label in labels])
    jql = f'filter = {filter_id} AND statusCategory != Done AND ({label_condition}) AND ((sprint is EMPTY AND "Story Points[Number]" is EMPTY) OR (sprint is not EMPTY AND "Story Points[Number]" is not EMPTY AND created <= -60d))'

    issues = call_jira_api(jira_domain, username, api_token, jql, max_results)['issues']

    # Group by project and limit to top 6 per project
    issues_by_project = {}
    for issue in issues:
        project_key = issue['fields']['project']['key']
        issues_by_project.setdefault(project_key, []).append(issue)
    issues_by_project = {k: v[:6] for k, v in issues_by_project.items()}

    rows = []
    for project_issues in issues_by_project.values():
        for issue in project_issues:
            created_date = datetime.strptime(issue['fields']['created'], '%Y-%m-%dT%H:%M:%S.%f%z')
            formatted_date = created_date.strftime('%m/%d/%Y %H:%M:%S')
            rows.append([
                issue['key'],
                issue['fields']['summary'],
                issue['fields']['assignee']['displayName'] if issue['fields'].get('assignee') else 'Unassigned',
                issue['fields']['reporter']['displayName'] if issue['fields'].get('reporter') else 'Unknown',
                issue['fields']['priority']['name'] if issue['fields'].get('priority') else 'N/A',
                issue['fields']['customfield_10000']['value'] if issue['fields'].get('customfield_10000') else 'N/A',
                formatted_date,
                ''  # Placeholder for Notes column
            ])

    if rows:
        sheet.append_rows(rows, value_input_option='USER_ENTERED')  # Append data, preserving formulas

# 6. Fetch and Write Tech Debt Tickets
def fetch_and_write_tech_debt_tickets(sheet, jira_domain, username, api_token, filter_id):
    jql = f'filter = {filter_id} AND statusCategory != Done AND ((sprint is EMPTY AND "Story Points[Number]" is EMPTY) OR (sprint is not EMPTY AND "Story Points[Number]" is not EMPTY AND created <= -60d)) ORDER BY created DESC'

    issue = call_jira_api(jira_domain, username, api_token, jql, 1)['issues'][0]

    created_date = datetime.strptime(issue['fields']['created'], '%Y-%m-%dT%H:%M:%S.%f%z')
    formatted_date = created_date.strftime('%m/%d/%Y %H:%M:%S')
    sheet.append_row([
        issue['key'],
        issue['fields']['summary'],
        issue['fields']['assignee']['displayName'] if issue['fields'].get('assignee') else 'Unassigned',
        issue['fields']['reporter']['displayName'] if issue['fields'].get('reporter') else 'Unknown',
        issue['fields']['priority']['name'] if issue['fields'].get('priority') else 'N/A',
        issue['fields']['customfield_10000']['value'] if issue['fields'].get('customfield_10000') else 'N/A',
        formatted_date,
        ''  # Placeholder for Notes column
    ], value_input_option='USER_ENTERED')  # Append data, preserving formulas

# 7. Call Jira API
def call_jira_api(jira_domain, username, api_token, jql, max_results):
    url = f'{jira_domain}/rest/api/2/search?jql={jql}&maxResults={max_results}'
    auth = (username, api_token)
    headers = {'Accept': 'application/json'}

    response = requests.get(url, auth=auth, headers=headers)
    response.raise_for_status()  # Raise an exception for bad status codes
    return response.json()

# 8. Main Execution
if __name__ == "__main__":
    fetch_and_write_tickets(new_sheet, jira_domain, username, api_token, general_filter_id, labels, 150)
    fetch_and_write_tech_debt_tickets(new_sheet, jira_domain, username, api_token, tech_debt_filter_id)